### DISCUSSION TITLE
How to: map `hyper::Body` to `lambda_http::Body`

### DISCUSSION THREAD

**Discussion by metasim** [aws/aws-lambda-rust-runtime#456] [Q&A] [answered]
I'm trying to create some impedance-matching glue between `lambda_http` and `axum`, so I can use the `Route` construct. I'm doing this via a `Service` trait, transforming an `axum::http::Response` into a `lambda_http::Response`. The place where I'm stuck is  transforming the `hyper::Body` (an opaque struct that behaves as a stream), into the `lambda_http::Body`, an Enum with owned content.

I've tried using `hyper::body::to_bytes`, but the `Response` from `axum` doesn't fulfill the `HttpBody` type bound.

I'm fairly new to Rust, so I may be "using it wrong", or trying to do something that should be done some other way. Basically would like to use `lambda_http` with `axum`.

**Comment by metasim** [comment, marked as answer]
I think I cracked the nut (but welcome feedback if there's a better way!):

```rust
use lambda_http as LH;
use axum::http as AH;
use axum::body::BoxBody;
use hyper::body::to_bytes;

async fn lambda_http_response_from_axum(resp: AH::Response<BoxBody>) -> Result<LH::Response<LH::Body>, LH::Error> {
    let (head, body) = resp.into_parts();
    let raw = to_bytes(body).await?;
    let resp = LH::Response::from_parts(head, LH::Body::Binary(raw.to_vec()));
    Ok(resp)
}
```

Still need to properly handle content type....

**Reply by bnusunny** [reply]
I did similar conversion in Lambda Adapter. It converts reqwest::Response to lambda_http::Body. It handles binary data in the same way as Application Load Balancer. The code is here. 

https://github.com/aws-samples/aws-lambda-adapter/blob/40b3b305f07e015de3b860ca23f01eb029a20337/src/main.rs#L90

**Reply by metasim** [reply]
Cool project @bnusunny! Thanks for the link!
